File: king-watchdog.py
"""
king-watchdog.py — keep all govOS services alive + Gordon AI coordinator.

Monitors:
  - Docker containers (studio-assets, neo4j, auth)
  - king-bridge (port 8109)
  - board-api (port 8799)
  - studio-assets (port 8108)
  - Ollama (port 11434)
  - Gordon AI coordinator (background)

Restarts anything that goes down. Logs to watchdog.log.
Run: python king-watchdog.py
"""
import json
import os
import subprocess
import sys
import threading
import time
import urllib.request
from datetime import datetime
from pathlib import Path

HERE    = Path(__file__).resolve().parent
LOG     = HERE / 'watchdog.log'
POLL_S  = 30  # check every 30 seconds
HST     = -10 * 3600  # Hawaii Standard Time offset

def now():
    return datetime.utcnow().strftime('%Y-%m-%d %H:%M UTC')

def log(msg):
    line = f'[{now()}] {msg}'
    print(line, flush=True)
    with open(LOG, 'a', encoding='utf-8') as f:
        f.write(line + '\n')

def http_ready(url, timeout=4):
    try:
        with urllib.request.urlopen(url, timeout=timeout) as r:
            return r.status == 200
    except Exception:
        return False

def docker_running(name):
    try:
        out = subprocess.check_output(
            ['docker', 'inspect', '--format', '{{.State.Status}}', name],
            stderr=subprocess.DEVNULL, text=True
        ).strip()
        return out == 'running'
    except Exception:
        return False

def docker_restart(name):
    log(f'RESTART docker:{name}')
    subprocess.run(['docker', 'restart', name], capture_output=True)

# ── Service definitions ───────────────────────────────────────────────────────
DOCKER_SERVICES = [
    {'name': '12sgi-king-auth-1',             'health': 'http://localhost:8101/api/v2/ready'},
    {'name': '12sgi-king-board-api-1',        'health': 'http://127.0.0.1:8799/health'},
    {'name': '12sgi-king-king-bridge-1',      'health': 'http://localhost:8109/api/v2/ready'},
]

EXTERNAL_HEALTH_CHECKS = [
    {'label': 'studio-assets', 'url': 'http://localhost:8108/api/v2/health'},
    {'label': 'ollama',        'url': 'http://localhost:11434/api/tags'},
]

# Processes we manage (cmd, ready_url, label)
_procs = {}

def ensure_process(label, cmd, ready_url, cwd=HERE):
    p = _procs.get(label)
    alive = p and p.poll() is None
    if alive and (not ready_url or http_ready(ready_url)):
        return  # all good
    if alive and ready_url and not http_ready(ready_url):
        log(f'UNRESPONSIVE {label} — killing and restarting')
        p.terminate()
        time.sleep(2)
        alive = False
    if not alive:
        log(f'START {label}: {" ".join(cmd)}')
        _procs[label] = subprocess.Popen(
            cmd, cwd=str(cwd),
            stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL
        )
        time.sleep(3)

# All services now run as Docker containers (docker-compose.v2.yml).
# Watchdog health-checks them but does NOT start them as subprocesses.
PROCESS_SERVICES = []

def check_docker():
    for svc in DOCKER_SERVICES:
        name = svc['name']
        if not docker_running(name):
            log(f'DOWN docker:{name}')
            docker_restart(name)
            time.sleep(5)
        elif svc.get('health') and not http_ready(svc['health']):
            log(f'UNHEALTHY docker:{name} — restarting')
            docker_restart(name)
            time.sleep(5)

def check_processes():
    for svc in PROCESS_SERVICES:
        ensure_process(svc['label'], svc['cmd'], svc.get('ready_url'))

def check_external():
    for chk in EXTERNAL_HEALTH_CHECKS:
        if not http_ready(chk['url'], timeout=3):
            log(f"WARNING: {chk['label']} not responding at {chk['url']}")

def tailscale_serve_setup():
    """Set up Tailscale serve rules for the king stack."""
    try:
        out = subprocess.check_output(['tailscale', 'serve', 'status'], text=True, stderr=subprocess.DEVNULL)
        serves = {
            '8109': 'king-bridge (API + AI)',
            '8799': 'board-api (owner console)',
        }
        for port, desc in serves.items():
            if port not in out:
                log(f'SETUP REQUIRED: run `tailscale serve --bg http://{port}` for {desc}')
            else:
                log(f'OK: Tailscale serve {port} ({desc})')
    except Exception as e:
        log(f'WARNING: Tailscale not available or error checking serve status: {e}')

def start_gordon_coordinator():
    """Start Gordon AI coordinator in a background thread."""
    try:
        from services.gordon_coordinator import GordonCoordinator
        coordinator = GordonCoordinator()
        log('STARTING: Gordon AI coordinator (background daemon)')
        thread = threading.Thread(
            target=coordinator.run_continuous_loop,
            kwargs={'interval': 30, 'duration_hours': None},
            daemon=True
        )
        thread.start()
        return coordinator
    except ImportError as e:
        log(f'WARNING: Gordon coordinator not available: {e}')
        return None
    except Exception as e:
        log(f'ERROR starting Gordon coordinator: {e}')
        return None

def main():
    log('=== king-watchdog started ===')
    log(f'Monitoring: {len(DOCKER_SERVICES)} Docker services + {len(EXTERNAL_HEALTH_CHECKS)} external endpoints')
    tailscale_serve_setup()
    
    # Start Gordon AI coordinator
    gordon = start_gordon_coordinator()

    # Initial check
    check_docker()
    check_processes()
    check_external()

    while True:
        try:
            check_docker()
            check_processes()
            check_external()
            time.sleep(POLL_S)
        except KeyboardInterrupt:
            log('watchdog stopped by user')
            for p in _procs.values():
                try: p.terminate()
                except: pass
            break
        except Exception as e:
            log(f'ERROR in watchdog loop: {e}')
            time.sleep(POLL_S)

if __name__ == '__main__':
    main()

File: king-watchdog.js
/**
 * king-watchdog.js — keep all services alive (13 tenants, 25 characters, 6 lipsync skills)
 *
 * Monitors:
 *   - Docker containers (studio-assets, neo4j, auth, etc.)
 *   - king-bridge API (port 8109)
 *   - Ollama (port 11434) — model inference
 *   - Static file server (port 8888)
 *
 * Features:
 *   - Auto-restart dead services
 *   - HTTP health checks
 *   - Graceful shutdown
 *   - Structured logging to watchdog.log
 *
 * Usage:
 *   node king-watchdog.js
 */

const fs = require('fs');
const path = require('path');
const http = require('http');
const { spawn, spawnSync } = require('child_process');

const HERE = __dirname;
const LOG_FILE = path.join(HERE, 'watchdog.log');
const POLL_MS = 30000;  // 30 seconds

// ─── Logging ─────────────────────────────────────────────────────────────────
function now() {
  return new Date().toISOString().slice(0, 19) + ' UTC';
}

function log(msg) {
  const line = `[${now()}] ${msg}`;
  console.log(line);
  fs.appendFileSync(LOG_FILE, line + '\n', { encoding: 'utf-8' });
}

// ─── HTTP Health Checks ──────────────────────────────────────────────────────
function httpReady(url, timeoutMs = 4000) {
  return new Promise((resolve) => {
    const timeoutHandle = setTimeout(() => resolve(false), timeoutMs);
    http.get(url, (res) => {
      clearTimeout(timeoutHandle);
      resolve(res.statusCode === 200);
    }).on('error', () => {
      clearTimeout(timeoutHandle);
      resolve(false);
    });
  });
}

// ─── Docker Operations ────────────────────────────────────────────────────────
function dockerRunning(containerName) {
  try {
    const result = spawnSync('docker', [
      'inspect',
      '--format', '{{.State.Status}}',
      containerName
    ], { encoding: 'utf-8' });
    return result.status === 0 && result.stdout.trim() === 'running';
  } catch (e) {
    return false;
  }
}

function dockerRestart(containerName) {
  log(`RESTART docker:${containerName}`);
  try {
    spawnSync('docker', ['restart', containerName]);
  } catch (e) {
    log(`ERROR restarting ${containerName}: ${e.message}`);
  }
}

// ─── Service Definitions ─────────────────────────────────────────────────────
const DOCKER_SERVICES = [\n  { name: 'studio-assets-studio-assets-1', health: 'http://localhost:8108/api/v2/ready' },\n  { name: 'studio-assets-studio-neo4j-1',  health: null },\n  { name: '12sgi-king-auth-1',             health: 'http://localhost:8101/api/v2/ready' },\n];\n\nconst PROCESS_SERVICES = [\n  {\n    id: 'king-bridge',\n    cmd: 'python',\n    args: [\n      '-m', 'uvicorn',\n      'services.king_bridge.app.main:app',\n      '--host', '127.0.0.1',\n      '--port', '8109',\n    ],\n    cwd: HERE,\n    healthUrl: 'http://localhost:8109/api/v2/ready',\n  },\n  {\n    id: 'static-server',\n    cmd: 'python',\n    args: [\n      '-m', 'http.server',\n      '8888',\n      '--directory', HERE,\n    ],\n    cwd: HERE,\n    healthUrl: 'http://localhost:8888/',\n  },\n];\n\n// ─── Process Management ──────────────────────────────────────────────────────\nconst procMap = new Map();\n\nfunction ensureProcess(svc) {\n  let proc = procMap.get(svc.id);\n  const alive = proc && !proc.killed;\n\n  if (alive && svc.healthUrl) {\n    return httpReady(svc.healthUrl).then((ready) => {\n      if (!ready) {\n        log(`UNRESPONSIVE ${svc.id} — killing and restarting`);\n        proc.kill();\n        procMap.delete(svc.id);\n        return startProcess(svc);\n      }\n    });\n  }\n\n  if (!alive) {\n    return startProcess(svc);\n  }\n}\n\nfunction startProcess(svc) {\n  return new Promise((resolve) => {\n    log(`START ${svc.id}: ${svc.cmd} ${svc.args.join(' ')}`);\n    const proc = spawn(svc.cmd, svc.args, {\n      cwd: svc.cwd,\n      stdio: ['ignore', 'ignore', 'ignore'],\n    });\n    procMap.set(svc.id, proc);\n    proc.on('error', (err) => {\n      log(`ERROR starting ${svc.id}: ${err.message}`);\n    });\n    // Wait 3s for startup\n    setTimeout(resolve, 3000);\n  });\n}\n\n// ─── Service Checks ──────────────────────────────────────────────────────────\nasync function checkDocker() {\n  for (const svc of DOCKER_SERVICES) {\n    if (!dockerRunning(svc.name)) {\n      log(`DOWN docker:${svc.name}`);\n      dockerRestart(svc.name);\n      await new Promise((r) => setTimeout(r, 5000));\n    } else if (svc.health && !(await httpReady(svc.health))) {\n      log(`UNHEALTHY docker:${svc.name} — restarting`);\n      dockerRestart(svc.name);\n      await new Promise((r) => setTimeout(r, 5000));\n    }\n  }\n}\n\nasync function checkProcesses() {\n  for (const svc of PROCESS_SERVICES) {\n    await ensureProcess(svc);\n  }\n}\n\nasync function checkOllama() {\n  const ready = await httpReady('http://localhost:11434/api/tags', 3000);\n  if (!ready) {\n    log('WARNING: Ollama not responding on :11434 — degraded inference mode');\n  }\n}\n\nfunction tailscaleServeHint() {\n  try {\n    const result = spawnSync('tailscale', ['serve', 'status'], {\n      encoding: 'utf-8',\n      stdio: ['ignore', 'pipe', 'ignore'],\n    });\n    if (result.status === 0) {\n      if (!result.stdout.includes('8109')) {\n        log('HINT: run `tailscale serve --bg http://8109` to expose king-bridge');\n      }\n      if (!result.stdout.includes('8888')) {\n        log('HINT: run `tailscale serve --bg http://8888` to expose static pages');\n      }\n    }\n  } catch (e) {\n    // tailscale not installed\n  }\n}\n\n// ─── Main Loop ───────────────────────────────────────────────────────────────\nasync function main() {\n  log('=== king-watchdog started ===');\n  log(`Monitoring: ${DOCKER_SERVICES.length} Docker + ${PROCESS_SERVICES.length} processes`);\n  log('Tenants: 13 (9 films, 1 game, 2 music videos, 1 civic studio)');\n  log('Characters: 25 • Lipsync skills: 6 • Render styles: 5 • Civic divisions: 12');\n  tailscaleServeHint();\n\n  // Initial startup\n  await checkDocker();\n  await checkProcesses();\n  await checkOllama();\n\n  let loopCount = 0;\n  while (true) {\n    try {\n      await checkDocker();\n      await checkProcesses();\n      // Every 5 minutes also check Ollama\n      if (loopCount % 10 === 0) {\n        await checkOllama();\n      }\n      loopCount++;\n      await new Promise((r) => setTimeout(r, POLL_MS));\n    } catch (err) {\n      log(`ERROR in watchdog loop: ${err.message}`);\n      await new Promise((r) => setTimeout(r, POLL_MS));\n    }\n  }\n}\n\n// ─── Graceful Shutdown ───────────────────────────────────────────────────────\nprocess.on('SIGINT', () => {\n  log('watchdog shutting down...');\n  for (const proc of procMap.values()) {\n    try {\n      proc.kill();\n    } catch (e) {}\n  }\n  process.exit(0);\n});\n\nif (require.main === module) {\n  main().catch((err) => {\n    log(`FATAL: ${err.message}`);\n    process.exit(1);\n  });\n}\n\nmodule.exports = { httpReady, dockerRunning, dockerRestart };\n